### PR TITLE
Support for multiple instructions 

### PR DESCRIPTION
**Pull request by IAvecilla**
This PR contains the implementation and tests for the following instructions:
- `xor`
- `nand`
- `or`
- `nor`

### CODE REVIEW

File: src/instructions/nor.rs [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+        .values()
+        .collect::<Vec<&CircuitIOType>>()
+        .as_slice()
+    {
+        [SimpleBoolean(left_operand), SimpleBoolean(right_operand)] => {
+            let a = AllocatedBool::new_witness(constraint_system.clone(), || left_operand.value())?;
+            let b = AllocatedBool::new_witness(constraint_system, || right_operand.value())?;
+            let result = AllocatedBool::nor(&a, &b)?;
+            let result = Boolean::from(result);
+            Ok(SimpleBoolean(result))
```
Comment (igamigo) [inline src/instructions/nor.rs:26]: Is there any specific reason why we need to allocate new witnesses here? 
Or could we do something like this: `let result = left_operand.or(right_operand)?.not();`

I don't mind either way but there might be something I'm missing

Comment (ilitteri) [reply src/instructions/nor.rs:26]: I don't think that it is necessary, if I'm not mistaken the allocation could be skipped and `a` and `b` could be the operand values.

Comment (jrchatruc) [reply src/instructions/nor.rs:26]: I imagine this has to do with the fact that the operands are `Boolean`s but `nor` is only implemented for `AllocatedBool`s. I think the correct thing to do here is to `match` the three possible `Boolean` values: `Is(AllocatedBool`, `Not(AllocatedBool)` and `Constant`. If they are allocated bools you use the internal value to perform `nor`, if they are constants you just do the `nor` without even generating constraints.

Comment (IAvecilla) [reply src/instructions/nor.rs]: > I imagine this has to do with the fact that the operands are `Boolean`s but `nor` is only implemented for `AllocatedBool`s. I think the correct thing to do here is to `match` the three possible `Boolean` values: `Is(AllocatedBool`, `Not(AllocatedBool)` and `Constant`. If they are allocated bools you use the internal value to perform `nor`, if they are constants you just do the `nor` without even generating constraints.

Yes, that was exactly the reason why I used the `AllocatedBools`, I saw the `nor` implementation and I tried to use it but we can use the `or` and `not` to do the same with `Booleans` and it should be all right. I did that change so let me know if it's okay.

Comment (jrchatruc) [reply src/instructions/nor.rs]: Yup, looks good to me 👍

**Review by jrchatruc — APPROVED** [review]

**Pull request merged**